### PR TITLE
CIN-09434: Ignore deleted display columns when rendering a form

### PR DESCRIPTION
**Pull request by DAquilinaCinchy**
# Description

Author or reviewer, don’t forget the [guidelines](https://cinchy.visualstudio.com/Cinchy/_wiki/wikis/Cinchy.wiki/448/Pull-Requests).

-----

# Overview

In cases where a link column has been configured with a set of display columns (which is done at the column level) and one
 or more of those columns is unavailable, the app was throwing an uncaught exception and the field would break.

# Details

It was decided that the correct behaviour was to squash the error and fail gracefully, rendering without the failing display column(s). This was accomplished by parsing the error statement returned by the initial attempt (since we can't know ahead of time what is broken), parsing out the relevant column(s), and then reconstructing the original query without the failing column(s) included.

# Risk

Minimal. Since the flow of logic is exactly the same, the only concern would be something breaking in the thoroughly-tested retry logic

# Testing

To reproduce:

- create a link column on Table A targeting Table B, adding at least one display column from Table B
- delete one or more of the display columns you have set from Table B
- add the link column on Table A to a form as a link or multichoice link field
- prior to this fix, the error should appear in the console and the link field will not populate
- after this fix, the link field should populate as expected, without displaying any values from the deleted display column. Note that the failed request still happens and can be inspected, and the console will reflect this error accordingly

# Docs

## Release notes

If you have deleted one or more display columns associated with a link or multichoice link field on a form, that field will no longer fail to load.

## Change log

N/A

### CODE REVIEW

File: src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts [hunk 1]
```diff
@@ -111,39 +111,85 @@ export class DropdownDatasetService {
...
-    });
+        let displayIndex = 0;
 
-    const result = new DropdownDataset(optionArray);
+        do {
+          if (row[`display-${displayIndex}`]) {
+            label += `, ${row[`display-${displayIndex}`]}`;
+          }
+        }
+        while (++displayIndex);
```
Comment (cysonya) [inline src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts:134]: When does this while loop stop? Doesn't `++displayIndex` keep adding 1 to itself?

Comment (DAquilinaCinchy) [reply src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts:134]: Good catch, though I already fixed it

File: src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts [hunk 2]
```diff
@@ -111,39 +116,93 @@ export class DropdownDatasetService {
...
-    const result = new DropdownDataset(optionArray);
+        do {
+          if (row[`display-${displayIndex}`]) {
+            label += `, ${row[`display-${displayIndex}`]}`;
+          }
+          else {
+            break;
+          }
+        }
+        while (++displayIndex);
```
Comment (cysonya) [inline src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts:134]: Just a suggestion, I think writing the while loop like this makes it easier to understand when the loop stops running
```suggestion
while (row[`display-${displayIndex}`]) {
    label += `, ${row[`display-${displayIndex}`]}`;
    displayIndex++;
}
```

File: src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts [hunk 3]
```diff
@@ -111,39 +116,93 @@ export class DropdownDatasetService {
...
+          // Will match a column name in the form of `[COLUMN_NAME]`, resulting in an array with [`[COLUMN_NAME]`]
+          columnMatch = errorMessage.substring(startingIndex).match(/\[([^\[\]]+)]/g);
+
+          if (columnMatch?.length) {
+            startingIndex = errorMessage.indexOf(columnMatch[0]) + columnMatch[0].length;
 
+            displayColumnsFoundInErrorResponse.push(columnMatch[0].replace("[", "").replace("]", ""));
+          }
+        }
+        while (columnMatch);
```
Comment (cysonya) [inline src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts:166]: Same here

Comment (cysonya) [inline src/app/dynamic-forms/service/cinchy-dropdown-dataset/cinchy-dropdown-dataset.service.ts:166]:
```suggestion
while ((columnMatch = errorMessage.substring(startingIndex).match(/\[([^\[\]]+)]/g)) !== null) {
    if (columnMatch.length) {
        startingIndex = errorMessage.indexOf(columnMatch[0]) + columnMatch[0].length;
        displayColumnsFoundInErrorResponse.push(columnMatch[0].replace("[", "").replace("]", ""));
    }
}
```

**Review by cysonya — DISMISSED** [review]

**Review by cysonya — APPROVED** [review]

**Pull request merged**